<a href="https://colab.research.google.com/github/andtoo14/kamilimu_assignments/blob/andrew_python_libraries_/Python_libraries_1_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**SECTION ONE:**

1. What is the primary purpose of data manipulation in the data science workflow?
- To clean, transform and structure raw data into a usable format to allow accurate analysing, modelling and visualization.

2. Name three Python libraries that are used for data manipulation, besides Pandas.
- NumPy
- Dask
- PySpark

3. What are the key limitations of standard Python lists when performing numerical computations with large datasets, which NumPy addresses?
- Slower performance due to lack of vecortization.
- Inneficient memory usage.
- Lack of element-wise operations.
- No broadcasting support for operations across arrays.









In [7]:
# 4. Explain the core difference between how Python lists and NumPy arrays handle the addition of two lists/arrays.
import numpy as np

#Python lists: list1 + list2 → concatenates the lists
my_list = [1, 2] + [3, 4]
print(my_list)

#NumPy arrays: np.array([1, 2]) + np.array([3, 4]) → element-wise addition
np.array([1, 2]) + np.array([3, 4])

[1, 2, 3, 4]


array([4, 6])

5. What is an ndarray?
- the Numerical Python library that gives us a powerful object.

6. List the advantages of using NumPy arrays over Python lists for numerical computations.
- much faster
- more memory-efficient
- tailor-made for numerical computations

7. What are the two primary data structures that Pandas introduces for working with tabular data? Briefly describe each.
- Series: A one-dimensional labeled array.
- DataFrame: A two-dimensional labeled data structure.

8. How does Pandas simplify working with tabular data compared to what NumPy primarily offers?
- Load data from CSVs, Excel, JSON, SQL, etc.
- Clean and preprocess the data
- Filter, sort, and transform columns
- Combine multiple datasets together
- Summarize data with groupings
- Export cleaned data for visualization or modeling

In [9]:
# 9. Name two ways to access a specific column in a Pandas DataFrame. Provide a simple example for each.
import pandas as pd

data = {
    'Name': ['Alice', 'Bob', 'Charlie'],
    'Age': [25, 30, 35],
    'Score': [85.5, 90.2, 78.9]
}

df0 = pd.DataFrame(data)
df1 = pd.DataFrame(data, index=['a.lice', 'b.ob', 'c.harlie'])

print(df0['Name'], "\n")  # bracket notation
print(df1.Age)  # dot notation

0      Alice
1        Bob
2    Charlie
Name: Name, dtype: object 

a.lice      25
b.ob        30
c.harlie    35
Name: Age, dtype: int64


10. What does df.isnull().sum() tell you about a DataFrame in Pandas, and why is this function important in data cleaning?
- It returns the total count of missing (null) values per column in a DataFrame. Its importance:
  - Helps identify incomplete or corrupted data
  - Guides decisions on data cleaning strategies (e.g., dropping or imputing missing values)


**SECTION TWO:**

In [10]:
import pandas as pd

df = pd.read_csv("https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv", sep='\t')

In [11]:
#Display the last 7 rows of the DataFrame df to get an initial look at the data.
df.tail(7)

order_id  quantity            item_name  \
4615      1832         1   Chicken Soft Tacos   
4616      1832         1  Chips and Guacamole   
4617      1833         1        Steak Burrito   
4618      1833         1        Steak Burrito   
4619      1834         1   Chicken Salad Bowl   
4620      1834         1   Chicken Salad Bowl   
4621      1834         1   Chicken Salad Bowl   

                                     choice_description item_price  
4615   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]     $8.75   
4616                                                NaN     $4.45   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75

In [12]:
#Print a summary of the DataFrame including the index dtype and column dtypes, non-null values, and memory usage. Comment on what you see.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


- The dataset has 4622 rows and 5 columns.
- All columns have non-null values — no missing data.
- The item_price column is of type object (string), which we’ll need to convert for numerical operations.
- Memory usage is relatively small.

In [13]:
#Determine how many observations have missing item|_price in the dataset.
df['item_price'].isnull().sum()

np.int64(0)

In [14]:
#Print the names of all columns in the DataFrame.
df.columns.tolist()

['order_id', 'quantity', 'item_name', 'choice_description', 'item_price']

In [15]:
#Identify which item_name appears most frequently in the orders.
df['item_name'].value_counts().idxmax()

'Chicken Bowl'

In [16]:
#Determine how many distinct item_name values are present in the dataset.
df['item_name'].nunique()

50

In [18]:
#Create a new DataFrame containing only the rows where the item_name is 'Chicken Bowl'.
chicken_bowl_df = df[df['item_name'] == 'Chicken Bowl']
chicken_bowl_df.head()

order_id  quantity     item_name  \
4          2         2  Chicken Bowl   
5          3         1  Chicken Bowl   
13         7         1  Chicken Bowl   
19        10         1  Chicken Bowl   
26        13         1  Chicken Bowl   

                                   choice_description item_price  
4   [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5   [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
13  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...    $11.25   
19  [Tomatillo Red Chili Salsa, [Fajita Vegetables...     $8.75   
26  [Roasted Chili Corn Salsa (Medium), [Pinto Bea...     $8.49

In [19]:
#The 'item_price' column is currently a string (e.g., '$2.39'). Convert it to a float type. You'll need to remove the dollar sign and then convert the data type.
df['item_price'] = df['item_price'].str.replace('$', '').astype(float)
df.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price  
0                                                NaN        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                                NaN        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98

In [20]:
#After converting 'item_price' to a numerical type, calculate the average item_price across all orders.
df['item_price'].mean()

np.float64(7.464335785374297)

In [21]:
#Determine the total number of unique orders in the dataset based on the order_id column.
df['order_id'].nunique()

1834

**SECTION THREE:**

1. Total Revenue Calculation (Pandas):

In [22]:
# Ensure item_price is a float
df['item_price'] = df['item_price'].replace('[\$,]', '', regex=True).astype(float)

# Calculate line item total
df['line_item_total'] = df['quantity'] * df['item_price']

# Calculate total revenue
total_revenue = df['line_item_total'].sum()
total_revenue

np.float64(39237.02)

2. Calculate Total Sales per Item (PySpark):

In [36]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, expr
from pyspark import SparkContext

# Start a Spark session
spark = SparkSession.builder.appName("Chipotle").getOrCreate()

# Convert pandas DataFrame to PySpark
spark_df = spark.createDataFrame(df)

# Convert item_price to float (if not already)
spark_df = spark_df.withColumn("item_price", col("item_price").cast("float"))

# Calculate line_item_total
spark_df = spark_df.withColumn("line_item_total", col("quantity") * col("item_price"))

# Group by item_name and sum line_item_total
sales_per_item = spark_df.groupBy("item_name").sum("line_item_total")\
                         .withColumnRenamed("sum(line_item_total)", "total_sales")

# Sort by total_sales descending and show top 5
top5_items = sales_per_item.orderBy(col("total_sales").desc()).limit(5)
top5_items.show()

Py4JJavaError: An error occurred while calling None.org.apache.spark.api.java.JavaSparkContext.
: org.apache.spark.SparkException: Only one SparkContext should be running in this JVM (see SPARK-2243).The currently running SparkContext was created at:
org.apache.spark.api.java.JavaSparkContext.<init>(JavaSparkContext.scala:58)
java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:247)
py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
py4j.Gateway.invoke(Gateway.java:238)
py4j.commands.ConstructorCommand.invokeConstructor(ConstructorCommand.java:80)
py4j.commands.ConstructorCommand.execute(ConstructorCommand.java:69)
py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
py4j.ClientServerConnection.run(ClientServerConnection.java:106)
java.base/java.lang.Thread.run(Thread.java:829)
	at org.apache.spark.SparkContext$.$anonfun$assertNoOtherContextIsRunning$2(SparkContext.scala:2845)
	at scala.Option.foreach(Option.scala:407)
	at org.apache.spark.SparkContext$.assertNoOtherContextIsRunning(SparkContext.scala:2842)
	at org.apache.spark.SparkContext$.markPartiallyConstructed(SparkContext.scala:2932)
	at org.apache.spark.SparkContext.<init>(SparkContext.scala:99)
	at org.apache.spark.api.java.JavaSparkContext.<init>(JavaSparkContext.scala:58)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:247)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
	at py4j.Gateway.invoke(Gateway.java:238)
	at py4j.commands.ConstructorCommand.invokeConstructor(ConstructorCommand.java:80)
	at py4j.commands.ConstructorCommand.execute(ConstructorCommand.java:69)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
	at java.base/java.lang.Thread.run(Thread.java:829)
